### ISSUE TITLE
Interfaces Deserialization for polymorph fields

### ISSUE THREAD

**Issue by jonathancregut** [arangodb/arangodb-java-driver#368]
First of all, thanks a lot for your amazing work.

I'm currently failing at getting one of my entities where I tried to do some polymorphism. 
I have a class, which has a field with an interface type (let's call it Vehicle). There can be different implementation for this field, Car, Bike,...

Serialization works great, however, when I try to get my document with the getDocument method I get an illegalArgumentException. 

This exception is happening in the UnsafeObjectFieldAccessorImpl.java set method 
```java
    public void set(Object obj, Object value)
        throws IllegalArgumentException, IllegalAccessException
    {
        ensureObj(obj);
        if (isFinal) {
            throwFinalFieldIllegalAccessException(value);
        }
        if (value != null) {
            if (!field.getType().isAssignableFrom(value.getClass())) {
                throwSetIllegalArgumentException(value);
            }
        }
        unsafe.putReference(obj, fieldOffset, value);
    }
```

Problem is, field.getType() yields my interface Vehicle, and value.getClass() yields my class Car or Bike. 

I don't get why Java is saying they are not assignable. 

Long story short, is the current implementation supposed to handle polymorphism as such ? 

If so, it's weird, the deserializer works really well, the hint field (_class) is used really well, the "Car" item itself is unserialized correctly, however the affectation of the Car to a Vehicle type field is failing.

FYI here is the stacktrace of where the error seems to be happening
![Capture d’écran 2020-11-18 à 21 11 03](https://user-images.githubusercontent.com/3193542/99582570-97240500-29e2-11eb-8ec4-c86d4003dc7d.png)

Best regards,

**Comment by rashtao** [comment]
@jonathancregut can you please provide a code snippet showing the failure?

For advanced use cases you can also use https://github.com/arangodb/jackson-dataformat-velocypack which allows you configuring polymorphic deserialization using Jackson databind API.

**Comment by jonathancregut** [comment]
@rashtao here is the failing case :

 
```java
@ToString
@Data
public class Action {
    private String name;
    private ActionConfiguration config;
}
```

The interface ActionConfiguration : 

```java
public interface ActionConfiguration {
    String getType();
}
```

And one of its implementation :

```java
@Data
public class ExternalActionConfiguration implements ActionConfiguration {
    
    private UUID uuid;


    @Override
    public String getType() {
        return "external";
    }
}
```


I deleted a lot of things but as a proof of concept I can reproduce the issue with this code.
Saving an Action instance works perfectly. 
However getting one fails on the error.

Is this the normal behavior and am I supposed to use Jackson deserializer in this scenario ?


Best regards.

**Comment by rashtao** [comment]
Your code works perfectly to me! Here is a simple test I wrote using it:
```java
    @Test
    public void roundTripTest() {
        Action action = new Action();
        action.setName("test");
        action.setConfig(new ExternalActionConfiguration());

        VPack vPack = new VPack.Builder().build();
        VPackSlice serialized = vPack.serialize(action);
        Action deserialized = vPack.deserialize(serialized, Action.class);
        assertThat(deserialized, is(action));
    }
```

and the json equivalent of the serialized form is:

```json
{
  "config": {
    "_class": "com.arangodb.velocypack.debug.PolymorphicDeserializationTest$ExternalActionConfiguration"
  },
  "name": "test"
}
```

**Comment by jonathancregut** [comment]
I tried in a unit test and indeed it does, I don't know why it's not working in the app. But that's my problem, at least it handles polymorphism perfectly. My bad then, sorry.

Thanks again for your help !
